### ISSUE TITLE
Drag/Drop causing errors while autosave is in-progress

### ISSUE THREAD

**Issue by arne-vdw** [spicywebau/craft-neo#823]
### Bug Description

Data from fields is removed when dragging-and-dropping while a draft autosave is in-progress.

https://github.com/spicywebau/craft-neo/assets/98316067/b101966f-be56-4a65-bae1-839dda407427

### Steps to reproduce

1. Add at least two Neo fields to a block.
2. Make a change to one of the fields.
3. While the entry is still auto-saving the draft, reorder the fields via drag-and-drop.
4. When the problem occurs with a block you will see the fields of another block within the current block for a moment

### Expected behaviour

Data should not be lost and the fields should be correctly reordered.

### Neo version

3.9.11

### Craft CMS version

4.5.12

### Does this issue involve templating, and if so, is eager-loading used?

This is not a templating issue

**Comment by svale** [comment]
Hi, 
we are seeing the same issue on a site recently upgraded to Craft 4. The bug seems pretty consistent in both production and dev environments. As a temporary workaround, we have asked the client to avoid drag'n drop if possible, but would very much like to present a better solution. Please let us know if we can help in any way to debug or test.

Versions:
Craft Pro 4.8.2
Neo	4.0.7

**Comment by svale** [comment]
I'd like to follow up on this as it is an ongoing issue for our client. We are seeing that content is lost on drag and drop. 

Here is a screen recording of a simple re-production of just two blocks where all the content is dropped when re-arranged .

https://github.com/spicywebau/craft-neo/assets/5339180/e944ce28-0054-4f3f-b32a-1dd610290613

Please let me know if we can share further details to help investigate.

Current versions
Craft Pro 4.8.8
Neo	4.1.2

**Comment by ttempleton** [comment]
@svale I continue to be unable to reproduce this. If possible, could you please send your composer files and database backup to plugins@spicyweb.com.au along with information about an entry where this is happening, and we'll have a look.

**Comment by svale** [comment]
Thank you for following up! Files and db are sent to plugins@spicyweb.com.au.

**Comment by ttempleton** [comment]
Thanks for sending those files @svale. I've made changes in the 4.x branch that I think have fixed this (or I'm having more trouble getting the timing right for it to happen).

Could you please run `composer require spicyweb/craft-neo:"4.x-dev as 4.1.2"` and let me know if you or your client experience this issue again?

**Comment by svale** [comment]
Hi @ttempleton , thank you for the update! The change is a clear improvement and it is now _mostly_ no issue when moving blocks. Unfortunately I can still trigger the error by moving re-ordering blocks  - especially when testing with an image block which at a random point "copies" the content from a text block:

Before             |  After  |  
:-------------------------:|:-------------------------:
![image](https://github.com/spicywebau/craft-neo/assets/5339180/96f8edd7-0db6-4dcc-b524-ef6d61a1e2e7)  |  ![image](https://github.com/spicywebau/craft-neo/assets/5339180/0f27be8f-ab5b-4b8e-8fc1-ef95fb4b1d76)

(See the block labeled "Bilder")

While the issue was pretty constant before updating, it is much more random now, so absolutely on the right track.

**Comment by myleshyson** [comment]
Ok we just noticed this happening as well. I'm having a hard time reproducing it when just using a rich text fields, but I can produce it consistently when I have a rich text field and some other type of field. You can see the video of what's happening below.

A couple things I'm noticing.
1. The autosave request fires while still holding a block.
2.  In the cases where this bug occurs, there's an autosave with the correct data, then one immediately after with blank data. 
3. It's not consistent when it happens, but if I drag certain blocks together it is consistent in happening at some point.

I wonder if pausing autosave while in a drag state, and triggering it after dropping would be the move here? There's some weirdness happening during drag, and i'm not sure if it's helpful to actually autosave before the user drops the block. 

Pretty sure something like this will work. When I do this manually (just in chrome dev tools), I don't experience the issue anymore.

```js
let editor = $('#main-form').data('elementEditor')

// on drag start
editor.pause()

// when dropped
editor.resume()

// force autosave
editor.checkForm(true)
```


https://github.com/spicywebau/craft-neo/assets/11747187/c3435aa8-276f-4b82-8875-6681dc58d8e3

**Comment by ttempleton** [comment]
Do the changes in the 4.x branch resolve the issue for you?

**Comment by myleshyson** [comment]
@ttempleton  Sweeeet. Yea so far that 4.x branch fixes the issue for me. There is a little UI jump that always caused the blocks to clear out, and now that jump doesn't clear the blocks anymore (you can see the jump I'm talking about in the video I posted).

**Comment by myleshyson** [comment]
Oops. Just deleted a comment saying I couldn't reproduce it, but I actually can. It's a little harder now but it still happens. There's a little UI jump that occurs when the data clears (you can see the jump I'm talking about in the video)

**Comment by myleshyson** [comment]
I was trying to open a pull request but I'm having trouble getting the local env running. I'm using Orbstack instead of Docker for Mac so I think there's maybe some weirdness there.

Anyway, the change I was gonna suggest is this. Working for me locally. It also seems to fix the UI jump I was talking about.

```js
// Input.js
this._blockSort = new BlockSort({
  // ...A bunch of stuff
  onDragStart: () => {
    this.$form.data('elementEditor')?.pause()
  },
  onDragStop: () => {
    this.$form.data('elementEditor')?.resume()
    this._updateBlockOrder()
    this._updateButtons()
  }
})
```

**Comment by ttempleton** [comment]
I've just made those changes in the 4.x branch, as well as some other changes to make sure the updating of visible layout elements in Neo blocks still happens.

@myleshyson @svale could you please confirm whether everything now works as expected with the current state of the 4.x branch?

**Comment by myleshyson** [comment]
Ok yea that fixes it for me. Tried for a couple minutes to reproduce and could not reproduce the losing data bug. There is still this weird jump that happens when dropping a block in a new position, but it's a minor UI bug. 


https://github.com/spicywebau/craft-neo/assets/11747187/6c788b1e-18af-496d-9319-46a0cb2240ec

**Comment by ttempleton** [comment]
Thanks for that feedback @myleshyson.

I've just made some other changes to how and when visible layout elements are updated, to fix a separate bug (condition rules for parent block field values not always being applied), but I think it could have also fixed that bug in your video. Could you please run those latest changes in the 4.x branch and let me know if that's fixed it?

**Comment by svale** [comment]
Thank you for digging into this, @ttempleton . I'm also seeing an improvement in stability, but unfortunately I can still reproduce:

https://github.com/spicywebau/craft-neo/assets/5339180/59e7713f-027d-4ea4-8a36-93dd3b54daab

See text and image is lost.

(In the screen recording console log, ignore "'setTimeout' handler took <N>ms", but  I'm not sure if the redactor.js event listener warning could be relevant ? Probably not as I can also reproduce without any text-blocks )  
 
 Neo:  4.x-dev (b4aa37d)
 Craft: 4.9.2

**Comment by ttempleton** [comment]
Thanks @svale. Do you notice any difference in stability when swapping two previously existing blocks, compared with one or both of the blocks being new?

**Comment by harry2909** [comment]
Hey just to add to the ongoing findings, we're also getting the above with content missing when moving blocks, however our client has also reported blocks disappearing entirely when saving as a new entry.

Their steps to reproduce were:
- Move a block around after cloning until content disappears.
- Save the entry as a new entry.
- Edit the new entry.
- Visit the original entry -> blocks have totally disappeared, date modified does not reflect the actual date modified (its older)

Appreciate this may be a craft bug, but just wanted to note it down to see if anyone else has had similar issues.

Thanks!

**Comment by ttempleton** [comment]
I suspect that some of the ongoing issues involve newly-created and unsaved blocks, so I've backported some changes from the 5.x branch to always load new blocks from the server, that have already been saved. I'm going to release this soon as version 4.2.0, regardless of whether it fixes this issue or not, but I'm hoping that it does fix this issue.

**Comment by ttempleton** [comment]
I've just released 4.2.0. I'm going to close this issue in hope that it has been fixed, but please let me know if the issues persist for you after upgrading.

**Comment by joepagan** [comment]
Hi @ttempleton just got the issue on release Neo release `v4.20` which @harry2909 referenced here: https://github.com/spicywebau/craft-neo/issues/823#issuecomment-2126959634

When saving as a new entry, which has draft revisions, the original entry loses it's content.

Potentially happens after a specific job in the queue has completed so doesn't appear to happen immediately. Could be after the queue jobs: `Deleting old Neo blocks` or `Saving Neo block structures for duplicated elements`.

Manage to reproduce and capture the incident on a screen recording:

https://github.com/spicywebau/craft-neo/assets/2337910/1ffa663d-f668-4594-a485-96e5cff2e28d

Please can you take another look?
